jax/experimental/ode.py

```diff
@@ -1,354 +1,368 @@
+    g = onp.zeros_like(flat_x)
+    for i in range(dim):
+      d = onp.zeros_like(flat_x)
+      d[i] = eps
+      g[i] = (f(unravel(flat_x + d)) - f(unravel(flat_x - d))) / (2.0 * eps)
+    return g
 
-            # Compute effect of moving measurement time.
-            vjp_cur_t = np.dot(func(yt[i, :], t[i], *func_args), g[i, :])
-            time_vjp_list.append(vjp_cur_t)
-            vjp_t0 = vjp_t0 - vjp_cur_t
+  def f(y, t, arg1, arg2):
+    return -np.sqrt(t) - y + arg1 - np.mean((y + arg2)**2)
 
-            # Run augmented system backwards to the previous observation.
-            aug_y0 = np.hstack((yt[i, :], vjp_y, vjp_t0, vjp_args))
-            aug_ans = odeint(augmented_dynamics, aug_y0, np.stack([t[i], t[i - 1]]), (flat_args,))
-            _, vjp_y, vjp_t0, vjp_args = unpack(aug_ans[1])
+  def onearg_odeint(args):
+    return np.sum(
+        odeint(f, args[2], args[0], args[1], atol=1e-8, rtol=1e-8))
 
-            # Add gradient from current output.
-            vjp_y = vjp_y + g[i - 1, :]
+  dim = 10
+  t0 = 0.1
+  t1 = 0.2
+  y0 = np.linspace(0.1, 0.9, dim)
+  fargs = (0.1, 0.2)
 
-        time_vjp_list.append(vjp_t0)
-        vjp_times = np.hstack(time_vjp_list)[::-1]
+  numerical_grad = nd(onearg_odeint, (y0, np.array([t0, t1]), fargs))
+  ys = odeint(f, fargs, y0, np.array([t0, t1]), atol=1e-8, rtol=1e-8)
+  ode_vjp = grad_odeint(f, fargs)
+  g = np.ones_like(ys)
+  exact_grad, _ = ravel_pytree(ode_vjp(g, ys, np.array([t0, t1])))
 
-        return None, vjp_y, vjp_times, unravel(vjp_args)
-    return vjp_all
+  assert np.allclose(numerical_grad, exact_grad)
 
 
-def nd(f, x, eps=0.0001):
-    flat_x, unravel = ravel_pytree(x)
-    D = len(flat_x)
-    g = onp.zeros_like(flat_x)
-    for i in range(D):
-        d = onp.zeros_like(flat_x)
-        d[i] = eps
-        g[i] = (f(unravel(flat_x + d)) - f(unravel(flat_x - d))) / (2.0 * eps)
-    return g
+if __name__ == '__main__':
 
+  test_grad_odeint()
 
-def test_odeint_vjp():
-    D = 10
-    t0 = 0.1
-    t1 = 0.2
-    y0 = np.linspace(0.1, 0.9, D)
-    fargs = (0.1, 0.2)
-    def f(y, t, arg1, arg2):
-        return -np.sqrt(t) - y + arg1 - np.mean((y + arg2)**2)
-
-    def onearg_odeint(args):
-        return np.sum(odeint(f, *args, atol=1e-8, rtol=1e-8))
-    numerical_grad = nd(onearg_odeint, (y0, np.array([t0, t1]), fargs))
-
-    ys = odeint(f, y0, np.array([t0, t1]), fargs, atol=1e-8, rtol=1e-8)
-    ode_vjp = grad_odeint(ys, f, y0, np.array([t0, t1]), fargs)
-    g = np.ones_like(ys)
-    exact_grad, _ = ravel_pytree(ode_vjp(g))
-
-    assert np.allclose(numerical_grad, exact_grad)
-
-
-if __name__ == "__main__":
-
-    def f(y, t, arg1, arg2):
-        return y - np.sin(t) - np.cos(t) * arg1 + arg2
-
-    t0 = 0.
-    t1 = 5.0
-    ts = np.linspace(t0, t1, 100)
-    y0 = np.array([1.])
-    fargs = (1.0, 0.0)
-
-    ys, (t_evals, y_evals, f_evals) =\
-        odeint(f, y0, ts, fargs, atol=0.001, rtol=0.001, return_evals=True)
-
-    # Set up figure.
```

Comment:
Could we keep this example+plotting code in the file somewhere? That way it'll be easy to factor out into an example file.